### PR TITLE
new task for updating annotations w/ rdc data 

### PR DESCRIPTION
**Pull request by jklugherz**
- new task `UpdateVariantAnnotationsTableWithUpdatedReferenceDataset` becomes a dependency of `UpdateVariantAnnotationsTableWithNewSamplesTask`
- `UpdatedReferenceDatasetCollectionTask` becomes a dependency of `BaseVariantAnnotationsTableTask`

### CODE REVIEW

File: v03_pipeline/lib/tasks/base/base_variant_annotations_table_test.py [hunk 1]
```diff
@@ -49,7 +51,11 @@ def setUp(self) -> None:
             ),
         )
 
-    def test_should_create_initialized_table(self) -> None:
+    @patch(
+        'v03_pipeline.lib.tasks.base.base_variant_annotations_table.UpdatedReferenceDatasetCollectionTask',
+    )
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/base/base_variant_annotations_table_test.py:56]: I was iffy on mocking the entire output of this task but since it's already unit tested in its own test file it made sense to me to just mock the output as complete instead of mocking all of the dataset tables, the config dict, and other task dependencies when it's not really the purpose of the test.

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 2]
```diff
@@ -83,63 +73,82 @@ def read_annotation_dependencies(self):
...
+                        self.callset_paths,
+                        self.project_guids,
+                        self.project_remap_paths,
+                        self.project_pedigree_paths,
+                    )
+                ],
+            )
+        upstream_table_tasks.extend(
+            [
+                UpdateVariantAnnotationsTableWithUpdatedReferenceDataset(
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:139]: this seems right to me, but on some local test runs I'm getting 
```
RuntimeError: Unfulfilled dependency at run time: UpdateVariantAnnotationsTableWithUpdatedReferenceDataset_SNV_INDEL_COMBINED_GRCh38_e9bde72efb
```
so I think the way I'm initializing these dependencies could be wrong

File: v03_pipeline/lib/tasks/base/base_variant_annotations_table.py [hunk 3]
```diff
@@ -7,10 +9,30 @@
...
 from v03_pipeline.lib.tasks.base.base_update_task import BaseUpdateTask
-from v03_pipeline.lib.tasks.files import GCSorLocalTarget, HailTableTask
+from v03_pipeline.lib.tasks.files import GCSorLocalTarget
+from v03_pipeline.lib.tasks.reference_data.updated_reference_dataset_collection import (
+    UpdatedReferenceDatasetCollectionTask,
+)
 
 
 class BaseVariantAnnotationsTableTask(BaseUpdateTask):
+    @cached_property
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/base/base_variant_annotations_table.py:19]: does `cached_property` make sense here? in my new task, we're reading the annotations table at least twice if the annotations table needs to be updated (in complete() and update_table())

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 4]
```diff
@@ -231,7 +244,7 @@ def update_table(self, ht: hl.Table) -> hl.Table:
         ):
             if rdc.requires_annotation:
                 continue
-            rdc_ht = annotation_dependencies[f'{rdc.value}_ht']
+            rdc_ht = self.rdc_annotation_dependencies[f'{rdc.value}_ht']
             new_variants_ht = new_variants_ht.join(rdc_ht, 'left')
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:248]: unsure if I need to keep this step, the left join seems significant

File: v03_pipeline/lib/annotations/mito.py [hunk 5]
```diff
@@ -79,9 +79,13 @@ def gt_stats(
         ref_samples_length = row.ref_samples[project_guid].length()
         heteroplasmic_samples_length = row.heteroplasmic_samples[project_guid].length()
         homoplasmic_samples_length = row.homoplasmic_samples[project_guid].length()
-        AC_het += heteroplasmic_samples_length # noqa: N806
-        AC_hom += homoplasmic_samples_length # noqa: N806
-        AN += (ref_samples_length + heteroplasmic_samples_length + homoplasmic_samples_length) # noqa: N806
+        AC_het += heteroplasmic_samples_length  # noqa: N806
```
Comment (bpblanken) [inline v03_pipeline/lib/annotations/mito.py:82]: Have our `ruff` settings diverged somehow? If so.. can we put these lint changes on another pr?

File: v03_pipeline/lib/reference_data/compare_globals.py [hunk 6]
```diff
@@ -15,131 +19,125 @@
 logger = logging.getLogger(__name__)
 
 
+def struct_to_dict(struct: hl.Struct) -> dict:
```
Comment (bpblanken) [inline v03_pipeline/lib/reference_data/compare_globals.py:22]: does something like `dict(hl.Struct(a=1, b=2))` work here?

File: v03_pipeline/lib/reference_data/compare_globals.py [hunk 7]
```diff
@@ -15,131 +19,125 @@
...
+        return cls(paths, versions, enums, selects)
+
+    @classmethod
+    def from_ht(
+        cls,
+        ht: hl.Table,
+        rdc: ReferenceDatasetCollection,
+        dataset_type: DatasetType,
+    ):
+        rdc_globals_struct = hl.eval(ht.index_globals())
```
Comment (bpblanken) [inline v03_pipeline/lib/reference_data/compare_globals.py:72]: I think you maybe only need `hl.eval(ht.globals)` here, or `ht.globals.collect()[0]`.

File: v03_pipeline/lib/tasks/base/base_variant_annotations_table.py (hunk 3 above)
Comment (bpblanken) [reply v03_pipeline/lib/tasks/base/base_variant_annotations_table.py:19]: I think the idea is correct!  We can definitely leave it, however hail is lazy, so I generally haven't bothered caching `hl.read_table`... it's mostly just initializing the table in memory and not doing any actual meaningful work.

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 8]
```diff
@@ -83,63 +73,82 @@ def read_annotation_dependencies(self):
             annotation_dependencies['gencode_mapping'] = hl.literal(
                 load_gencode(GENCODE_RELEASE, ''),
             )
-
         return annotation_dependencies
 
     def requires(self) -> list[luigi.Task]:
+        upstream_table_tasks = []
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:79]: You could maybe slot `UpdateVariantAnnotationsTableWithUpdatedReferenceDataset` here when initializing the list!

File: v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    BaseVariantAnnotationsTableTask,
+)
+
+
+class UpdateVariantAnnotationsTableWithUpdatedReferenceDataset(
+    BaseVariantAnnotationsTableTask,
+):
+    rdc = luigi.EnumParameter(enum=ReferenceDatasetCollection)
+
+    def complete(self) -> bool:
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:19]: Is this method gonna work correctly if the reference dataset collection has been correctly updated and persisted before this task is run?

File: v03_pipeline/lib/reference_data/compare_globals.py (hunk 6 above)
Comment (jklugherz) [reply v03_pipeline/lib/reference_data/compare_globals.py:22]: it does but if it's nested like this, 
```
g = hl.Struct(
                versions=hl.Struct(),
                paths=hl.Struct(
                    a='gs://a_path_OLD.ht',
                ),
                enums=hl.Struct(),
            )
print(dict(g))
```
prints {'versions': Struct(), 'paths': Struct(a='gs://a_path_OLD.ht'), 'enums': Struct()}, 

I can simplify the function a bit by converting the struct into a dict initially:
```
def struct_to_dict(struct: hl.Struct) -> dict:
    struct = dict(struct)
    for field in struct:
        if isinstance(struct[field], hl.Struct):
            struct[field] = struct_to_dict(struct[field])
    return struct
```

Comment (jklugherz) [reply v03_pipeline/lib/reference_data/compare_globals.py:22]: do you think it's just unnecessary to turn structs into the dicts? I think my reasoning for this was that it was easier for me to think about equality checks with just simple python data structures as opposed to hail structs

Comment (bpblanken) [reply v03_pipeline/lib/reference_data/compare_globals.py:22]: I'm not sure!  It looks like hail's types work just fine with `__eq__`:

```
In [5]: hl.Struct(a=1,b={'d': 2}) == hl.Struct(a=1,b={'d': 1})
Out[5]: False

In [6]: hl.Struct(a=1,b={'d': 1}) == hl.Struct(a=1,b={'d': 1})
Out[6]: True
```

File: v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py (hunk 9 above)
Comment (jklugherz) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:19]: not sure what you mean by this, got a sec before 5 to chat?

Comment (bpblanken) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:19]: I went through and re-read this before our chat so I could re-fresh myself... I think this is actually correct as is and I misunderstood it before!  sorry for the confusing question!

File: v03_pipeline/lib/tasks/base/base_variant_annotations_table.py (hunk 3 above)
Comment (jklugherz) [reply v03_pipeline/lib/tasks/base/base_variant_annotations_table.py]: makes sense, I just made it a property

File: v03_pipeline/var/test/tasks/mock_complete_task.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import luigi
+
+
+class MockCompleteTask(luigi.Task):
```
Comment (bpblanken) [inline v03_pipeline/var/test/tasks/mock_complete_task.py:4]: This probably belongs in `v03_pipeline/lib/test`.  `var` is meant for miscellaneous non-code files!

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 11]
```diff
@@ -246,38 +251,14 @@ def update_table(self, ht: hl.Table) -> hl.Table:
...
-                    **ht.globals.versions,
-                    **rdc_globals.versions,
-                ),
-                enums=hl.Struct(
-                    **ht.globals.enums,
-                    **rdc_globals.enums,
-                ),
-            )
-        ht = annotate_enums(ht, self.reference_genome, self.dataset_type)
+        ht = self.fix_globals(ht, clear_globals=True)
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:261]: Could we get rid of the `clear_globals` arg if we managed the `rdc_globals` in the method but left `annotate_enums` here?

File: v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+            self.rdc,
+            annotations_ht_globals,
+            rdc_ht_globals,
+            self.dataset_type,
+        )
+        return len(updated_datasets_for_rdc) == 0
+
+    def update_table(self, ht: hl.Table) -> hl.Table:
+        rdc_ht = self.rdc_annotation_dependencies[f'{self.rdc.value}_ht']
+        rdc_datasets = self.rdc.datasets(self.dataset_type)
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:43]: Do we want to touch every `dataset` here or just the ones that require an update?

Comment (bpblanken) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:43]: As in... do we want to use: 

```
rdc_datasets = get_datasets_to_update(
             self.rdc,
             annotations_ht_globals,
             rdc_ht_globals,
             self.dataset_type,
         )
```

Comment (jklugherz) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:43]: I think I implemented it the way I did because I thought I needed to join every dataset in the rdc - see line 49 `ht = ht.join(rdc_ht, 'left')`..

but can I do `ht = ht.join(rdc_ht[dataset], 'left')`? would it save time/computation to do a join for each dataset that needs an update?

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py (hunk 11 above)
Comment (jklugherz) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:261]:
```
        # 5) Fix up the globals.
        ht = ht.annotate_globals(
            paths=hl.Struct(),
            versions=hl.Struct(),
            enums=hl.Struct(),
        )
        ht = self.fix_globals(ht)
```
I realized this would probably work.

what's special about the enums in `annotate_enums`?

Comment (bpblanken) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:261]: Everything in `annotate_enums` is specific to fields that are annotated from the callset VCF and aren't coming from the reference data.  

It actually looks like we've introduced weirdness by slotting the clinvar enums into that file... which should be able to resolve by the upcoming work to make cached reference dataset query a task.

File: v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py (hunk 12 above)
Comment (bpblanken) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:43]: Yeah.  I think we only want to re-join against the field that needs it.  It _should_ reduce work!

File: v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    BaseVariantAnnotationsTableTask,
+):
+    rdc = luigi.EnumParameter(enum=ReferenceDatasetCollection)
+    _datasets_to_update: ClassVar[list[str]] = []
+
+    def complete(self) -> bool:
+        self._datasets_to_update.clear()
+
+        if not super().complete():
+            self._datasets_to_update.extend(
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:26]: This might be cleaner w/ just setting to the list.  It should always be empty at this point right?

```
self._datasets_to_update = self.rdc.datasets(self.dataset_type)
```

Comment (jklugherz) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:52]: we still need `extend` now that the task checks multiple rdcs

File: v03_pipeline/lib/model/reference_dataset_collection.py [hunk 14]
```diff
@@ -94,3 +94,40 @@ def for_reference_genome_dataset_type(
         if not Env.ACCESS_PRIVATE_REFERENCE_DATASETS:
             return [rdc for rdc in rdcs if rdc.access_control == AccessControl.PUBLIC]
         return rdcs
+
+    @classmethod
+    def for_dataset(cls, dataset: str) -> 'ReferenceDatasetCollection':
```
Comment (bpblanken) [inline v03_pipeline/lib/model/reference_dataset_collection.py:99]: It'd be cool if we could use the existing method so we don't have to duplicate lists.  It's basically just calling `datasets(dataset_type)` for all possible`DatasetType`?

File: v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py (hunk 13 above)
Comment (bpblanken) [reply v03_pipeline/lib/tasks/reference_data/update_variant_annotations_table_with_updated_reference_dataset.py:52]: What about on line 34?

File: v03_pipeline/lib/model/reference_dataset_collection.py (hunk 14 above)
Comment (jklugherz) [reply v03_pipeline/lib/model/reference_dataset_collection.py]: ok I cleaned this up to use `datasets` function, and I added `dataset_type` as an arg to `for_dataset` since we have it from the calling task.

**Review by bpblanken — APPROVED** [review]
:shipit:
